flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherImpl.java

```diff
@@ -148,97 +154,117 @@ private void fetchMetrics() {
-                                    retrieveAndQueryMetrics(queryServiceAddress);
-                                }
-                            }
-                        },
-                        executor);
-
-                // TODO: Once the old code has been ditched, remove the explicit TaskManager query
-                // service discovery
-                // TODO: and return it as part of requestMetricQueryServiceAddresses. Moreover,
-                // change the MetricStore such that
-                // TODO: we don't have to explicitly retain the valid TaskManagers, e.g. letting it
-                // be a cache with expiry time
-                CompletableFuture<Collection<Tuple2<ResourceID, String>>>
-                        taskManagerQueryServiceGatewaysFuture =
-                                leaderGateway.requestTaskManagerMetricQueryServiceAddresses(
-                                        timeout);
-
-                taskManagerQueryServiceGatewaysFuture.whenCompleteAsync(
-                        (Collection<Tuple2<ResourceID, String>> queryServiceGateways,
-                                Throwable throwable) -> {
-                            if (throwable != null) {
-                                LOG.debug(
-                                        "Requesting TaskManager's path for query services failed.",
-                                        throwable);
-                            } else {
-                                List<String> taskManagersToRetain =
-                                        queryServiceGateways.stream()
-                                                .map(
-                                                        (Tuple2<ResourceID, String> tuple) -> {
-                                                            queryServiceRetriever
-                                                                    .retrieveService(tuple.f1)
-                                                                    .thenAcceptAsync(
-                                                                            this::queryMetrics,
-                                                                            executor);
-                                                            return tuple.f0.getResourceIdString();
-                                                        })
-                                                .collect(Collectors.toList());
-
-                                metrics.retainTaskManagers(taskManagersToRetain);
-                            }
-                        },
-                        executor);
+                List<CompletableFuture<Void>> waitingMetricsFutures = new ArrayList<>();
+                waitingMetricsFutures.add(queryJmMetricsFuture(leaderGateway));
+                waitingMetricsFutures.add(queryTmMetricsFuture(leaderGateway));
+                return FutureUtils.waitForAll(waitingMetricsFutures);
```

Comment:
I'd suggest to add a `whenCompleteAsync` before returning the future, to add some debug logs if the future is completed exceptionally. This aligns with the previous behavior (the error stack should suggest which step has failed).